crates/pico_core/src/database.rs

```diff
@@ -1,17 +1,221 @@
+use std::any::Any;
+
+use dashmap::DashMap;
+use once_map::OnceMap;
+
 use crate::{
+    dependency_stack::DependencyStack,
     dyn_eq::DynEq,
     epoch::Epoch,
-    source::Source,
-    source::SourceId,
-    storage::{Storage, StorageMut},
+    generation::Generation,
+    index::Index,
+    key::Key,
+    node::{Dependency, DerivedNode, DerivedNodeId, DerivedNodeRevision, NodeKind, SourceNode},
+    params::ParamId,
+    source::{Source, SourceId},
 };
 
-pub trait Database {
-    fn storage(&self) -> &impl Storage<Self>;
-    fn storage_mut(&mut self) -> &mut impl StorageMut<Self>;
-    fn current_epoch(&self) -> Epoch;
-    fn increment_epoch(&mut self) -> Epoch;
-    fn get<T: Clone + 'static>(&mut self, id: SourceId<T>) -> T;
-    fn set<T: Source + DynEq>(&mut self, source: T) -> SourceId<T>;
-    fn remove<T>(&mut self, id: SourceId<T>);
+#[derive(Debug)]
+pub struct Database {
+    pub current_epoch: Epoch,
+    pub dependency_stack: DependencyStack<(Epoch, Dependency)>,
+    pub param_id_to_index: DashMap<ParamId, Index>,
+    pub derived_nodes: DerivedNodesStore,
+    pub source_nodes: DashMap<Key, SourceNode>,
+    pub capacity: usize,
+    pub epoch_to_generation_map: OnceMap<Epoch, Box<Generation>>,
+}
+
+impl Database {
+    pub fn new(capacity: usize) -> Self {
+        let epoch_to_generation_map = OnceMap::new();
+        epoch_to_generation_map.insert(Epoch::new(), |_| Box::new(Generation::new()));
```

Comment:
nit: probably worth making `Epoch::new` a variable for clarity, even though it implements copy